crates/parcel_core/src/plugin/transformer_plugin.rs

```diff
@@ -26,103 +25,70 @@ pub struct ResolveOptions {
 /// A function that enables transformers to resolve a dependency specifier
 pub type Resolve = dyn Fn(PathBuf, String, ResolveOptions) -> Result<PathBuf, anyhow::Error>;
 
-/// Compile a single asset, discover dependencies, or convert the asset to a different format
-///
-/// Many transformers are wrappers around other tools such as compilers and preprocessors, and are
-/// designed to integrate with Parcel.
+/// Transformers may run against:
 ///
-pub trait TransformerPlugin: Debug + Send + Sync {
-  /// Whether an AST from a previous transformer can be reused to prevent double-parsing
-  ///
-  /// This function should inspect the type and version of the AST to determine if it can be
-  /// reused. When the AST is reused, parsing is skipped. Otherwise the previous transformer
-  /// generate function is called, and the next transformer will parse that result.
-  ///
-  fn can_reuse_ast(&self, ast: AST) -> bool;
-
-  /// Parse the asset code into an AST
-  ///
-  /// This function is called when an AST is not available and can_reuse_ast returns false.
-  ///
-  fn parse(
-    &mut self,
-    config: &PluginConfig,
-    asset: &Asset,
-    resolve: &Resolve,
-  ) -> Result<AST, anyhow::Error>;
-
-  /// Transform the asset and/or add new assets
-  fn transform(
-    &mut self,
-    config: &PluginConfig,
-    asset: &mut Asset,
-    resolve: &Resolve,
-  ) -> Result<Vec<Asset>, anyhow::Error>;
-
-  // Perform processing after the transformation
-  fn post_process(
-    &mut self,
-    config: &PluginConfig,
-    assets: Vec<&Asset>,
-  ) -> Result<Vec<Asset>, anyhow::Error>;
-
-  /// Stringify the AST
-  ///
-  /// This function is called when the next transformer AST cannot be reused, or this is the last
-  /// transformer in a pipeline.
-  ///
-  fn generate(&self, asset: Asset, ast: AST) -> Result<GenerateOutput, anyhow::Error>;
+/// * File-paths that have just been discovered
+/// * Outputs of previous transformation steps, which are in-place modified
+/// * These two scenarios are distinguished
+pub enum TransformationInput {
+  FilePath(PathBuf),
```

Comment:
There's other info we'll need outside of the file path, e.g. sideEffects, isSource, env. Anything that's specific to this transform or derived from resolution. We can add this as need be, but maybe we should name the type clearer?

Reply:
We can add that as we actually consume it

Reply:
shouldn't the input of a transformer always be an Asset? Otherwise transformers would need to be aware of whether they are running first or later in a pipeline. In v2 transformers always receive an asset as input, and core constructs an the initial asset to pass to the first transformer.